src/shared/sleep-config.c

```diff
@@ -102,6 +107,179 @@ int parse_sleep_config(SleepConfig **ret_sleep_config) {
+               /* Handling case when no battery is present. Exception here file is manually deleted / corrupted */
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity: %m");
+
+        r = safe_atoi(bat_cap, &battery_capacity);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to parse battery capacity: %m");
+
+        if (battery_capacity < 0 || battery_capacity > 100)
+               return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "Invalid battery capacity");
+
+        return battery_capacity;
+}
+
+/* Read system and battery identifier from specific location and generate hash of it */
+static int get_system_battery_identifier_hash(uint8_t ret[static SHA256_DIGEST_SIZE]) {
+        _cleanup_free_ char *bat_manufacturer = NULL, *bat_model_name = NULL, *bat_serial_number = NULL;
+        sd_id128_t machine_id, product_id;
+        struct sha256_ctx hash;
+        int r;
+
+        sha256_init_ctx(&hash);
+        r = read_one_line_file("/sys/class/power_supply/BAT0/manufacturer", &bat_manufacturer);
+        if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/manufacturer is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/manufacturer: %m");
+        else if (strlen(bat_manufacturer) == 0)
+               log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "/sys/class/power_supply/BAT0/manufacturer is empty: %m");
+        else
+               sha256_process_bytes(bat_manufacturer, strlen(bat_manufacturer) + 1, &hash);
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/model_name", &bat_model_name);
+        if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/model_name is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/model_name: %m");
+        else if (strlen(bat_model_name) == 0)
+               log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "/sys/class/power_supply/BAT0/model_name is empty: %m");
+        else
+               sha256_process_bytes(bat_model_name, strlen(bat_model_name) + 1, &hash);
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/serial_number", &bat_serial_number);
+        if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/serial_number is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/serial_number: %m");
+        else if (strlen(bat_serial_number) == 0)
+               log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "/sys/class/power_supply/BAT0/serial_number is empty: %m");
+        else
+               sha256_process_bytes(bat_serial_number, strlen(bat_serial_number) + 1, &hash);
```

Comment:
you are basically have the same code here three times: read a file, and hash it. Please split out into a single function that takes the path to do this on as parameter.

Reply:
Makes sense. working on this.

Reply:
Fixed